main/src/main/scala/sbt/internal/server/LanguageServerProtocol.scala

```diff
@@ -135,6 +136,47 @@ private[sbt] object LanguageServerProtocol {
                 case NonFatal(e) =>
                   errorRespond("Cancel request failed")
               }
+            case r: JsonRpcRequestMessage if r.method == "textDocument/completion" =>
+              val param = Converter.fromJson[CompletionParams](json(r)).get
+
+              try {
+                Option(EvaluateTask.lastEvaluatedState.get) match {
```

Comment:
Probably I should've commented this on the cancellation PR too, but generally this class should be used as a glue code to interpret LSP context, and then quickly pass it along to NetworkChannel etc.
See for example:

```scala
             case r: JsonRpcRequestMessage if r.method == "sbt/exec" =>
               val param = Converter.fromJson[SbtExecParams](json(r)).get
               appendExec(Exec(param.commandLine, Some(r.id), Some(CommandSource(name))))
               ()
```

In the same fashion, I think the parameters needed for shell completion should be put into a pseudo case class, and sent to a callback method like `appendShellCompletion(ShellCompletion(...))`. That should avoid the globally holding on to the state.